DAQ/src/ArtBinaryPacketsFromDigis_module.cc

```diff
@@ -0,0 +1,1113 @@
+      pos += sizeof(TrackerDataPacket);
+    }
+
+    dataStream.back().second = pos;
+
+
+  }
+
+  void  ArtBinaryPacketsFromDigis::fillTrackerDMABlocks(raw_data_list_t& dataStream, tracker_data_block_list_t const& trkData) {
+
+    auto curDTCID = trkData.front().first.DTCID;
+    bool first = true;
+    if (_diagLevel > 1) {
+      std::cout << "[ArtBinaryPacketsFromDigis::fillTrackerDMABlocks] trkData.size() = " << trkData.size() << std::endl;
+    }
+    for (auto& dataBlock : trkData) {
+
+      fillTrackerDataStream(dataStream, dataBlock);
+
+      if (_diagLevel > 1) {
+	if (dataBlock.first.DTCID != curDTCID || first) {
+	  std::cout << "================================================" << std::endl;
+	  //std::cout << "\t\tTimestamp: " << ts << std::endl;
+	  std::cout << "\t\tDTCID: " << (int)dataBlock.first.DTCID << std::endl;
+	  std::cout << "\t\tSYSID: " << (int)dataBlock.first.SubsystemID << std::endl;
+	  curDTCID = dataBlock.first.DTCID;
+	  first = false;
+	}
+	if (dataBlock.first.PacketCount > 0) {
+	  printHeader(dataBlock.first);
+	  printTrackerData(dataBlock.second);
+	}
+      }
+
+
+    } // End loop over DataBlocks
+  }
+
+  //--------------------------------------------------------------------------------
+  //
+  //--------------------------------------------------------------------------------
+  void   ArtBinaryPacketsFromDigis::fillEmptyTrackerDataPacket(TrackerDataPacket& TrkData) {
+    bzero(&TrkData, sizeof(TrkData));
+  }
+
+
+
+  //--------------------------------------------------------------------------------  
+  // 
+  //--------------------------------------------------------------------------------
+  void   ArtBinaryPacketsFromDigis::fillEmptyHeaderDataPacket(DataBlockHeader& headerData, uint64_t& EventNum,
+							      uint8_t& ROCId, uint8_t& DTCId, uint8_t Subsys) {
+
+    // Fill in the byte count field of the header packet
+    // Word 0
+    headerData.ByteCount = sizeof(DataBlockHeader);
+    // Word 1
+    headerData.Hopcount = 0;//ask Eric!!!//FIX ME!
+    headerData.PacketType = 5;//PacketType::Dataheader; 
+
+    headerData.ROCID = ROCId;
+    headerData.unused1 = 0;//ask Eric!
+
+    headerData.SubsystemID = Subsys;//DTCLib::DTC_Subsystem_Tracker; //: 3;
+
+    headerData.Valid = 1;
+    // Word 2
+    headerData.PacketCount = 0;
+    headerData.unused2 = 0;// : 5;
+    // Word 3
+    uint64_t timestamp = EventNum;
+    headerData.TimestampLow = static_cast<adc_t>(timestamp & 0xFFFF);
+    // Word 4
+    headerData.TimestampMed = static_cast<adc_t>((timestamp >> 16) & 0xFFFF);
+    // Word 5
+    headerData.TimestampHigh = static_cast<adc_t>((timestamp >> 32) & 0xFFFF);
+    // Word 6
+    headerData.Status = 0; // 0 corresponds to "TimeStamp had valid data"
+    headerData.FormatVersion = format_version;
+    // Word 7
+    headerData.DTCID = DTCId;
+    uint8_t  evbMode = 0;//maybe off-spill vs on-spill?
+    headerData.EVBMode = evbMode;
+  }
```

Comment:
Use symbolic names for the numeric literals.  Which of these are used in other modules/functions?  Those should come from a common, authoritative source.

Reply:
Hi Rob,

This is on our to-do list. We didn't decide yet with Eric how to set up "EventMode" class in mu2e_artdaq. I was waiting to have some code working for the CRV online reconstruction to start working on it.

Reply:
Are you saying the when we have real data, you won't need to fill headerData since artdaq will have already filled it for you?  Or something else?